src/App.js

```diff
@@ -522,7 +526,31 @@ export default class App extends React.Component<{}, AppState> {
         );
     }
 
+    componentDidMount() {
+        if (window.location.search != null) {
+            const query = new URLSearchParams(window.location.search);
+            const programQuery = query.get('program');
+            if (programQuery != null) {
+                try {
+                    this.setState({
+                        program: this.programSerializer.deserialize(programQuery)
+                    });
+                } catch(err) {
+                    console.log(`Error parsing program: ${programQuery}`);
+                    console.log(err.toString());
+                }
+            }
+        }
+    }
+
     componentDidUpdate(prevProps: {}, prevState: AppState) {
+        if (this.state.program !== prevState.program) {
+            const encodedURI = encodeURI(this.programSerializer.serialize(this.state.program));
```

Comment:
This is definitely the right general approach, but we should use `encodeURIComponent()` rather than `encodeURI()`. `encodeURIComponent()` encodes characters that have special meaning in URIs, such as "?", whereas `encodeURI()` does not. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent. I would also suggest renaming this variable from 'encodedURI' to 'encodedProgram'.